### PR TITLE
[BUGFIX] Fix pagination in DFG-Viewer GridView

### LINKED ISSUE #1101: Misbehavior in the GridView pagination
## Description

There is a major difference between **ListView** and **GridView** when it comes to pagination.

- In the **ListView**, one screen page always corresponds to one pagination page. This is the correct expected behaviour!

image

- In the DFG-Viewer **GridView**, however, one screen page (pagination page) corresponds to several document pages. This means that for the links to a pagination page (gridpage number [1,2,3...]), a distinction must be made between the page count of the pagination page [1,2,3...] and the document page [1,11,22...] to which the link must be positioned. This is always the first document page of a block of 10 document pages (or less at the last pagination page), assuming the standard of 10 document pages per screen / pagination page.

image

### There are some errors in the current implementation of the pagination of the gridview:

1. The links for the individual pagination pages [1,2,3,...] to the required document page [1,11,22,...] are incorrect
2. "previous page" link is incorrect
3. "next page" link ist incrorrect
4. "next page" link and "last page" link is set to deactivated too early
5. "last page" link is calculated incorrectly

## Reproduction

Steps to reproduce the behaviour:

1. Open a document in DFG-Viewer (with more then 50 pages)
2. Click on GridView
for error type 1: use a page button higher then 1
for error type 3 and 4: Use the "Next" button until the end or until it is deactivated
for error type 4: use "Last page" button
for error type 2: first use "next" button until you reach the page 3, then use the "previous" button
5. See error

### 1. The links for the individual pagination pages [1,2,3,...] to the required document page [1,11,22,...] are incorrect.
No distinction is currently made between the pagination page [1,2,3,….] and document pages [1,11,22,...] on a pagination page.

To correct this error, data must be provided for each loop pass for the respective document page [1,11,22,…] to be taken into account.

For the loop: `<f:for each="{pagination.pages}" as="page">`

```
// to calculate the values for generation the links for the pagination pages
$itemsPerPage = $paginator->getPublicItemsPerPage();

$pages = array();
// calculation of the values for the page links within the loop
foreach (range($firstPage, $lastPage) as $i) {
	// calculation of the startRecordNumber
	$startRecordNumber = $itemsPerPage * $i;
	$startRecordNumber = $startRecordNumber + 1;
	$startRecordNumber = $startRecordNumber - $itemsPerPage;

	// save last startRecordNumer for LastPage button
	$lastStartRecordNumberGrid = $startRecordNumber;
	// array with label as screen/pagination page number
	// and startRecordNumer for correct structure of the link
	//<f:link.action action="{action}" 
	//      addQueryString="true" 
	//      argumentsToBeExcludedFromQueryString="{0: 'tx_dlf[page]'}" 
	//      additionalParams="{'tx_dlf[page]': page.startRecordNumber}" 
	//      arguments="{searchParameter: lastSearch}">{page.label}</f:link.action>
	$pages[$i] = array('label' => $i, 'startRecordNumber' => $startRecordNumber);
};

$nextPageNumber = $pages[$aktPageNumber + 1][‚startRecordNumber‘];
$previousPageNumber = $pages[$aktPageNumber - 1][‚startRecordNumber‘];
```
#### calculate nextPageNumber, prevPage and lastPage

The variable pagination / the return value of the function buildSimplePagination in Classes/Controller/AbstractController.php must also return more values for the current page.

- lastPageNumber => last screen page [1,2,3,…]

- lastPageNumberG => Document page [1,11,22,….] to build the last screen page. This is the first document of the last block of 10 (or less at the last page) documents on the last screen page

- firstPageNumber => always 1

- nextPageNumber => Document page [1,11,22,….] to build the next screen page
- nextPageNumberG => Number of the screen page [1,2,3,…] for the next screen page

- previousPageNumber => Document page [1,11,22,….] to build up the previous screen page
- previousPageNumberG => Number of the screen page [1,2,3,…] for the previous screen page

- startRecordNumber => is, as far as I know, only needed within the loop for each page entry, but not directly within pagination

- endRecordNumber => is not required in any situation as far as I know

- currentPageNumber => Number of the current screen page [1,2,3,…]

- pages => array of numbers of the screen pages [1,2,3,…], for compatibility reasons

- pagesG => array with two keys

  - label	=> Number of the screen page [1,2,3,…]
  - startRecordNumber => First document page [1,11,22,….] of this block of 10 document pages on the same screen page [1,2,3,…]

```
return [
    'lastPageNumber' => $lastPage,
    'lastPageNumberG' => $lastStartRecordNumberGrid,
    'firstPageNumber' => $firstPage,
    'nextPageNumber' => $nextPageNumber,
    'nextPageNumberG' => $paginator->getCurrentPageNumber() + 1,
    'previousPageNumber' => $previousPageNumber,
    'previousPageNumberG' => $paginator->getCurrentPageNumber() - 1,
    'startRecordNumber' => $pagination->getStartRecordNumber(),
    'endRecordNumber' => $pagination->getEndRecordNumber(),
    'currentPageNumber' => $paginator->getCurrentPageNumber(),
    'pages' => range($firstPage, $lastPage),
    'pagesG' => $pages            
];
```
**Names of variables and array keys are only suggestions!**

### 2. "previous page" link is incorrect
The calculation for previousPageNumber is incorrect. It should be the startRecordNumber (document page [1,11,22,….]) of the previous pagination page. As a result, the generated link is incorrect.

Fix this with:
`$previousPageNumber = $pages[$aktPageNumber - 1]['startRecordNumber'];`

### 3. "next page" link is incorrect
The calculation for nextPageNumber is incorrect. It should be the startRecordNumber (document page [1,11,22,….]) of the next pagination page. As a result, the generated link is incorrect.

Fix this with:
`$nextPageNumber = $pages[$aktPageNumber + 1]['startRecordNumber'];`

### 4. The "next page" and "last page" button can only be used to scroll a few pages, it is set to deactivated too early because the calculation of the end is incorrect. The calculation is linked to the pagination pages and not to the, in this case higher,  document page.

For an example, I have used a document with 326 document pages. With 10 document pages in the GridView, this results in 33 GridView pages. With the "next" button you can scroll to pagination page 4, then the "next" button is deactivated.

With the new variables it’s possible to correct this:
`<f:if condition="{pagination.nextPageNumberG} && {pagination.nextPageNumberG} <= {pagination.lastPageNumber}">
`

### 5. "last page" link is calculated incorrectly

With the new variable it’s possible to correct this:

`<f:link.action action="{action}" addQueryString="true" argumentsToBeExcludedFromQueryString="{0: 'tx_dlf[page]'}" additionalParams="{'tx_dlf[page]': pagination.lastPageNumberG}" ...
`

from:
 
```
return [
    'lastPageNumber' => $lastPage,
    'lastPageNumberG' => $lastStartRecordNumberGrid,
    ....

```

## Additional Context

Some of the errors are also described in issue #1029

### PR DESCRIPTION
**Pull request by BFallert**
See Issue "Misbehavior in the GridView pagination" https://github.com/kitodo/kitodo-presentation/issues/1101. A detailed description can also be found there.

There are some errors in the current implementation of the pagination of the gridview (used in DFG-Viewer):

1. The links for the individual pagination pages [1,2,3,...] to the required document page [1,11,22,...] are incorrect
1. "previous page" link is incorrect
1. "next page" link ist incrorrect
1. "next page" link and "last page" link is set to deactivated too early
1. "last page" link is calculated incorrectly

These bugs will be fixed

### CODE REVIEW

**Comment by BFallert** [comment]
I get a error message from PHPStan Static Analysis
`Error: Call to an undefined method TYPO3\CMS\Core\Pagination\PaginatorInterface::getPublicItemsPerPage().`
but I don't know how to fix this error

**Comment by beatrycze-volk** [comment]
> I get a error message from PHPStan Static Analysis `Error: Call to an undefined method TYPO3\CMS\Core\Pagination\PaginatorInterface::getPublicItemsPerPage().` but I don't know how to fix this error

If you are sure that this function exists then you can add ignore for this error. 

I have looked into API and it doesn't look like it exists: https://api.typo3.org/11.5/interface_t_y_p_o3_1_1_c_m_s_1_1_core_1_1_pagination_1_1_paginator_interface.html

Public Member Functions

 	[withItemsPerPage]

 	[withCurrentPageNumber]

    [getPaginatedItems]

 	[getNumberOfPages]

 	[getCurrentPageNumber]

 	[getKeyOfFirstPaginatedItem]

 	[getKeyOfLastPaginatedItem]

Are you sure that it exists?

**Comment by BFallert** [comment]
The function is contained in the file Classes/Pagination/PageGridPaginator.php
inserted in https://github.com/kitodo/kitodo-presentation/commit/1660fb2c8a367d72e6f53fae6313165fcd289fd8 and customised by #1018 
Is this the right place? or should it only be ignored if it would exist in the API?

**Comment by beatrycze-volk** [comment]
I would suspect that in that case it can be added to ignore in PHPStan.

@sebastian-meyer what do you think?

**Comment by BFallert** [comment]
The last commit is based on an idea and code from @csidirop

File: Resources/Private/Partials/Lists/Pagination.html [hunk 1]
```diff
@@ -1,70 +1,69 @@
...
+            </li>
+            <li class="last disabled">
+                <span>{pagination.lastPageNumber}</span>
+            </li>
+        </f:else>
+    </f:if>
+</ul>
 
-</html>
+</html>
```
Comment (sebastian-meyer) [inline Resources/Private/Partials/Lists/Pagination.html]:
```suggestion
</html>

```

File: Classes/Controller/AbstractController.php [hunk 2]
```diff
@@ -350,16 +350,77 @@ protected function buildSimplePagination(PaginationInterface $pagination, Pagina
...
+            if (get_class($pagination) == 'TYPO3\CMS\Core\Pagination\SimplePagination') {  // ListView
+                $lastStartRecordNumberGrid = $i; // save last $startRecordNumber for LastPage button
+
+                $pages[$i] = [
+                    'label' => $i,
+                    'startRecordNumber' => $i
+                ];
+            } else { // GridView
+                // to calculate the values for generation the links for the pagination pages
+                $itemsPerPage = $paginator->getPublicItemsPerPage();
```
Comment (sebastian-meyer) [inline Classes/Controller/AbstractController.php:370]: This should hopefully fix the PHPStan error:
```suggestion
                /** @var \Kitodo\Dlf\Pagination\PageGridPaginator $paginator */
                $itemsPerPage = $paginator->getPublicItemsPerPage();
```

**Review by sebastian-meyer — CHANGES_REQUESTED** [review]

File: Classes/Controller/AbstractController.php (hunk 2 above)
Comment (BFallert) [reply Classes/Controller/AbstractController.php:370]: I have added your suggestion, but unfortunately it does not quite solve the problem. There is still an error message:

`PHPDoc tag @var for variable $paginator contains unknown class Kitodo\Dlf\Controller\Kitodo\Dlf\Pagination\PageGridPaginator.`

Comment (sebastian-meyer) [reply Classes/Controller/AbstractController.php:370]: My fault: The namespace should be absolute, i. e. with a prefixed backslash. I fixed my proposal above.

**Review by sebastian-meyer — APPROVED** [review]

**Pull request merged**